plugins/Files/js/sagas/files.js

```diff
@@ -0,0 +1,227 @@
+	}
+}
+
+// Set the user's renter allowance.
+function* setAllowanceSaga(action) {
+	try {
+		const response = yield siadCall('/renter')
+		const newAllowance = SiaAPI.siacoinsToHastings(action.funds).add(response.settings.allowance.funds)
+		yield siadCall({
+			url: '/renter',
+			method: 'POST',
+			qs: {
+				funds: newAllowance.toString(),
+				hosts: allowanceHosts,
+				period: allowancePeriod,
+			},
+		})
+		yield put(actions.setAllowanceCompleted())
+		yield put(actions.closeAllowanceDialog())
+	} catch (e) {
+		yield put(actions.setAllowanceCompleted())
+		yield put(actions.closeAllowanceDialog())
+		sendError(e)
+	}
+}
+
+// Calculate monthly storage cost from a requested monthly storage amount
+function* calculateStorageCostSaga(action) {
+	try {
+		const response = yield siadCall('/renter/hosts/active')
+		const cost = estimatedStoragePriceGBSC(response.hosts, action.size, allowancePeriod)
+		yield put(actions.setStorageCost(cost.round(3).toString()))
+		yield put(actions.setStorageSize(action.size))
+	} catch (e) {
+		console.error(e)
+		yield put(actions.setStorageSize(''))
+		yield put(actions.setStorageCost('0'))
+	}
+}
+
+// Query Siad for the current wallet balance.
+function* getWalletBalanceSaga() {
+	try {
+		const response = yield siadCall('/wallet')
+		const confirmedBalance = SiaAPI.hastingsToSiacoins(response.confirmedsiacoinbalance).round(2).toString()
+		yield put(actions.receiveWalletBalance(confirmedBalance))
+	} catch (e) {
+		sendError(e)
+	}
+}
+
+const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms))
+
+function* setAllowanceProgressBarSaga() {
+	try {
+		let response = yield siadCall('/renter/contracts')
+		if (!response.contracts) {
+			return
+		}
+		const initialContracts = response.contracts.length
+		while (true) {
+			response = yield siadCall('/renter/contracts')
+			const deltaContracts = response.contracts.length - initialContracts
+			if (deltaContracts === allowanceHosts) {
+				break
+			}
+			const progress = Math.floor((deltaContracts / allowanceHosts) * 100)
+			yield put(actions.setAllowanceProgress(progress.toString()))
+			yield delay(500)
+		}
+	} catch (e) {
+		sendError(e)
+	}
+}
+
+function* uploadFileSaga(action) {
+	try {
+		const filename = Path.basename(action.source)
+		const destpath = action.siapath + filename
+		yield siadCall({
+			url: '/renter/upload/' + destpath,
+			method: 'POST',
+			qs: {
+				source: action.source,
+			},
+		})
+	} catch (e) {
+		sendError(e)
+	}
+}
+
+// Recursively upload the folder at the directory `source`
+function *uploadFolderSaga(action) {
+	try {
+		const files = readdirRecursive(action.source)
+		const folderName = Path.basename(action.source)
+		const siafiles = files.map((file) => ({
+			siapath: file.substring(file.indexOf(folderName), file.lastIndexOf(Path.basename(file))),
+			source: file,
+		}))
+		for (let i = 0; i < siafiles.size; i++) {
+			yield put(actions.uploadFile(siafiles.get(i).siapath, siafiles.get(i).source))
```

Comment:
dang, that's annoying. Still, you should be able to loop over the values instead of the index, right? Even if you have to use `.toList`, I think it's cleaner.